### PR TITLE
CBPB Manure Moduel: Methane and Carbon Emissions

### PR DESCRIPTION
**Pull request by michaelRichards99**
This PR adds the calculations for methane and carbon emissions, as specified in the Compost Bed Pack Barn (CBPB) [design doc](https://docs.google.com/document/d/1TIhbR919SXGFkIQDVzEDwxCJjopx2fkBtW_w5-TuJ00/edit?usp=sharing).

The entry point for methane emissions is `calc_ifsm_methane_emission()`.
The entry point for carbon emissions is `calc_total_carbon_decomposition()`
Both located in `gas_emissions.py`

### CODE REVIEW

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 1]
```diff
@@ -647,3 +648,177 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+
+    @classmethod
+    def calc_mcf(cls, ambient_barn_temp: float) -> float:
+        """
+        Calculate the Methane Conversion Factor (MCF) using the exponential function:
+        MCF(T) = MCF_CONSTANT_A * e^(MCF_CONSTANT_B * T)
+
+        Parameters
+        ----------
+        ambient_barn_temperature : float
```
Comment (allisterakun) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py]:
```suggestion
        ambient_barn_temp : float
```
Keep it consistent with the function definition?

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 2]
```diff
@@ -647,3 +648,177 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+        """Calculates emission of methane for a day using an adaptation of the tier 2 approach
+        of the IPCC(2006), given ambient barn temperature and a methane conversion factor for the manure
+        management.
+
+        Parameters
+        ----------
+        manure_volatile_solids : float
+            The volatile solids (in kg)
+        ambient_barn_temperature : float
+            The ambient barn temperature (in Celsius)
```
Comment (allisterakun) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py]:
```suggestion
        manure_volatile_solids : float
            The volatile solids (in kg)
            
        ambient_barn_temperature : float
            The ambient barn temperature (in Celsius)
```
Consistent docstring formatting?

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 3]
```diff
@@ -647,3 +648,177 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+    def calc_ifsm_methane_emission(cls, manure_volatile_solids: float, ambient_barn_temp: float) -> float:
+        """Calculates emission of methane for a day using an adaptation of the tier 2 approach
+        of the IPCC(2006), given ambient barn temperature and a methane conversion factor for the manure
+        management.
+
+        Parameters
+        ----------
+        manure_volatile_solids : float
+            The volatile solids (in kg)
+        ambient_barn_temperature : float
```
Comment (allisterakun) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py]:
```suggestion
        ambient_barn_temp : float
```
Consistency with function definition.

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 4]
```diff
@@ -647,3 +648,177 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+        temperature : float
+            The temperature of the medium (in Celsius)
+
+        Returns
+        -------
+        float
+            The microbial decomposition rate per day (unitless)
+
+        """
+        return ManureConstants.EFFECTIVE_MICROBIAL_DECOMP_RATE * \
```
Comment (allisterakun) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py:1162]: M.1.B.4
image

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 5]
```diff
@@ -647,3 +648,177 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+        -------
+        float
+            The calculated methane emissions (in kg) for the given ambient barn temperature.
+
+        """
+        if manure_volatile_solids < 0:
+            raise ValueError("Mass must be positive.")
+        methane_conversion_factor = GasEmissions.calc_mcf(ambient_barn_temp)
+        methane_emissions_in_kg = (manure_volatile_solids * GasEmissionConstants.Bo *
+                                   GasEmissionConstants.METHANE_FACTOR * methane_conversion_factor) / 100
```
Comment (allisterakun) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py]: M.1.A.1
image

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 6]
```diff
@@ -647,3 +648,177 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+        ambient_barn_temperature : float
+            The ambient barn temperature (in Celsius).
+
+        Returns
+        -------
+        float
+            The calculated Methane Conversion Factor (MCF) for the given ambient barn temperature.
+
+        """
+        return GasEmissionConstants.MCF_CONSTANT_A * math.exp(GasEmissionConstants.MCF_CONSTANT_B * ambient_barn_temp)
```
Comment (allisterakun) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py:1112]: M.1.A.2
image

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 7]
```diff
@@ -647,3 +648,177 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+
+        max_microbial_decom_rate = cls._microbial_decomp_rate(decomposition_temp)
+        slow_decomp_rate = cls._microbial_decomp_rate(compost_bed_pack_temp)
+        exponent_coeff = decay * (days_since_last_tillage - lag)
+
+        c_decomp_rate = (
+            (max_microbial_decom_rate - slow_decomp_rate)
+            * math.exp(exponent_coeff)
+            * slow_decomp_rate
+        )
```
Comment (allisterakun) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py:1200]: M.1.B.2
image

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 8]
```diff
@@ -647,3 +648,177 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+
+        """
+        if not (0.0 < oxygen_mole_fraction < 1.0):
+            raise ValueError(f"{oxygen_mole_fraction=} must be in the range [0, 1]")
+        if not (0.0 < oxygen_ambient_air_mole_fraction < 1.0):
+            raise ValueError(f"{oxygen_ambient_air_mole_fraction=} must be in the range [0, 1]")
+        anaerobic_coefficient = (
+            (oxygen_mole_fraction / (oxygen_half_saturation_constant + oxygen_mole_fraction))
+            * ((oxygen_half_saturation_constant + oxygen_ambient_air_mole_fraction) / oxygen_ambient_air_mole_fraction)
+        )
```
Comment (allisterakun) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py]: M.1.B.3
image

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 9]
```diff
@@ -647,3 +648,177 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+
+
+        Parameters
+        ----------
+        oxygen_mole_fraction : float
+            Mole fraction of oxygen in the air within the windrow
+        oxygen_half_saturation_constant : float
+            half saturation constant for oxygen gas
+        oxygen_ambient_air_mole_fraction : fot
+            mole fraction of oxygen gas in ambient air
```
Comment (allisterakun) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py]:
```suggestion
        oxygen_mole_fraction : float
            Mole fraction of oxygen in the air within the windrow
            
        oxygen_half_saturation_constant : float
            half saturation constant for oxygen gas
            
        oxygen_ambient_air_mole_fraction : fot
            mole fraction of oxygen gas in ambient air
```

**Review by allisterakun — APPROVED** [review]
Also, the docstring position and format for the three constants files are not consistent with each other.

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py (hunk 4 above)
Comment (allisterakun) [reply RUFAS/routines/manure/gas_emissions/gas_emissions.py:1162]: I just put these here for reference.

**Comment by michaelRichards99** [comment]
Thanks for the feedback, Allister. The existing doc-strings in these files are using an older format than what we're using. There are plans to update these eventually to use the numpy style. For now, any new code in these files will use the numpy style.

**Review by varrma — APPROVED** [review]
Looks Good Michael !!

File: RUFAS/routines/manure/constants/bedding_constants.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class BeddingConstants:
+    DEFAULT_CARBON_AVAILABLE_IN_BEDDING = 0.35
+    """
+    Default proportion of carbon available in bedding, (unitless, [0, 1]). Default is set to 0.35.
+    """
```
Comment (ltp23) [inline RUFAS/routines/manure/constants/bedding_constants.py]: I think we should move this constant to one of the other two constants files.

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 11]
```diff
@@ -647,3 +648,207 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
         MS = GasEmissionConstants.FRACTION_OF_HANDLED_MANURE
         MF = GasEmissionConstants.METHANE_FACTOR
         return Bo * MCF * MS * MF * manure_volatile_solids
+
+    @classmethod
+    def calc_mcf(cls, ambient_barn_temp: float) -> float:
```
Comment (ltp23) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py]: Maybe changing mcf to its full name? I don't know if this is somewhat related to the preexisting MCF constant (0.79) that we have.

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 12]
```diff
@@ -647,3 +648,207 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+            The ambient barn temperature (in Celsius)
+
+        Returns
+        -------
+        float
+            The calculated methane emissions (in kg) for the given ambient barn temperature.
+
+        """
+        if manure_volatile_solids < 0:
+            raise ValueError("Mass must be positive.")
```
Comment (ltp23) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py]: I think the error message should mention volatile solids specifically.

File: RUFAS/routines/manure/gas_emissions/gas_emissions.py [hunk 13]
```diff
@@ -647,3 +648,207 @@ def calc_methane_emission_from_anaerobic_lagoon(cls, manure_volatile_solids: flo
...
+
+        c_decomp_rate = (
+            (max_microbial_decom_rate - slow_decomp_rate)
+            * math.exp(exponent_coeff)
+            * slow_decomp_rate
+        )
+        return c_decomp_rate
+
+    @classmethod
+    def _anaerobic_effect(
```
Comment (ltp23) [inline RUFAS/routines/manure/gas_emissions/gas_emissions.py]: I would add _calc to the name.

File: tests/test_manure_module/test_gas_emissions.py [hunk 14]
```diff
@@ -48,6 +48,109 @@ def test_calc_E_CH4_slurry_storage(is_enclosed: bool, mocker: MockerFixture) ->
...
+def test_carbon_decomposition_rate(mocker: MockerFixture, days_since_last_tillage: int, lag: int) -> None:
+    """Tests _carbon_decomposition_rate() in gas_emissions.py."""
+
+    # Arrange
+    max_decomp_rate = 2.0
+    min_decomp_rate = 1.0
+    _ = mocker.patch(
+        'RUFAS.routines.manure.gas_emissions.gas_emissions.GasEmissions._microbial_decomp_rate',
+        side_effect=[max_decomp_rate, min_decomp_rate]
+    )
```
Comment (ltp23) [inline tests/test_manure_module/test_gas_emissions.py]: Not sure about the use of the unnamed variable here. It's ok to leave it out, or you can give it a name and assert the call counts.

**Review by ltp23 — APPROVED** [review]
Thanks, Michael. The code looks good to me.

**Pull request merged**